class.c

```diff
@@ -908,6 +916,8 @@ rb_define_module_id_under(VALUE outer, ID id)
 		     " (%"PRIsVALUE")",
 		     outer, rb_id2str(id), rb_obj_class(module));
 	}
+        /* Module may have been defined in Ruby and not pin-rooted */
+        rb_gc_register_mark_object(module);
```

Comment:
By the way, `rb_vm_add_root_module()` now seems to serve exactly the same purpose as `rb_gc_register_mark_object()`. They both root and pin. Maybe we can delete `rb_vm_add_root_module()`.

Reply:
I think we can.  AFAIK the array of arrays is to prevent bloating (but I am just guessing).  I think it would be better to use that than an `st_table`.